GVFS/GVFS.Common/NuGetUpgrader/NuGetUpgrader.cs

```diff
@@ -0,0 +1,368 @@
+﻿using GVFS.Common.FileSystem;
+using GVFS.Common.Git;
+using GVFS.Common.Tracing;
+using NuGet.Protocol.Core.Types;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Threading;
+
+namespace GVFS.Common
+{
+    public class NuGetUpgrader : IProductUpgrader

```

Comment:
I think this is fair feedback on the high level design, and is something that @kewillford and I also discussed. I think the `IProductUpgrader` interface we have now is an evolution from where the code started from, and it enables us to meet our primary goal (enable upgrade with different upgrade mechanism). I think we could decompose this further - but is a tradeoff between extra layers of abstraction and implementation. As you note - I think the class would pretty cleanly separate into two more focused classes.

I will think about this some more and at least do a bit more to see exactly what this would look like.

Reply:
I am going to resolve this conversation for now. I think what we have now, while it could be tweaked further, is good for our current goals. I will mark this a potential future improvement, though.